### PR TITLE
Custom nav2 behavior tree

### PR DESCRIPTION
**Pull request by PhileasL**
# Purpose

The purpose of this PR is for the robot to automatically choose what movement to do considering the goal and its current pose (position / orientation).

The robot have to move **linearly and as much perpendicular to the wall as possible in specific area** (by specific area i mean near walls and in ports).

# How ?

For this need, we'll need 2 things :
- Another controller but slower and more accurate,
- Reshaping the navigation behavior tree.

# Another controller

Well, TEB controller is pretty cool, why not reuse the same but slower and more accurate ? That's what have been done.
The second controller is shaped for displacement in specific areas, in can get as close to obstacles as 0.5 cm.

# Reshaping the navigation behavior tree

The [readme](https://github.com/robotique-ecam/cdfr/blob/de025a084878fff4ca53378fef71960fe11cd920/src/navigation/bt_custom_plugins/README.md) from the bt_custom_plugins describes itself pretty well.

### CODE REVIEW

File: src/navigation/bt_custom_plugins/bt_split_goal/include/split_goal_action.hpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
+#ifndef SPLIT_GOAL_ACTION_HPP_
+#define SPLIT_GOAL_ACTION_HPP_
+
+#include <memory>
+#include <string>
+
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav_msgs/msg/path.hpp"
```
Comment (PhileasL) [inline src/navigation/bt_custom_plugins/bt_split_goal/include/split_goal_action.hpp:8]: Dependencies.

File: src/navigation/bt_custom_plugins/bt_split_goal/package.xml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  <exec_depend>builtin_interfaces</exec_depend>
+  <exec_depend>geometry_msgs</exec_depend>
+  <exec_depend>sensor_msgs</exec_depend>
+  <exec_depend>nav2_msgs</exec_depend>
+  <exec_depend>nav_msgs</exec_depend>
+  <exec_depend>tf2</exec_depend>
+  <exec_depend>tf2_ros</exec_depend>
+  <exec_depend>tf2_geometry_msgs</exec_depend>
+  <exec_depend>nav2_util</exec_depend>
+  <exec_depend>lifecycle_msgs</exec_depend>
```
Comment (PhileasL) [inline src/navigation/bt_custom_plugins/bt_split_goal/package.xml:47]: Dependencies.

File: src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
+#include <string>
+#include <memory>
+#include <limits>
+
+#include "nav_msgs/msg/path.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "behaviortree_cpp_v3/decorator_node.h"
```
Comment (PhileasL) [inline src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp:7]: Dependencies.

File: src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  }
+  return 10;
+}
+
+void SplitGoal::setAutoQuaternions(){
+  geometry_msgs::msg::Quaternion q;
+
+  if (get_out_need_){
+    geometry_msgs::msg::Quaternion q_out = current_pose_.pose.orientation;
+    RCLCPP_WARN(node_->get_logger(), "q_out z: %f, w: %f", q_out.z, q_out.w);
```
Comment (PhileasL) [inline src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp:107]: Remove debugging log.

File: src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        || (abs(q_out.z + 1.0)<0.22 &&  abs(q_out.w)<0.22) ){
+      q.z = 1.0;
+      q.w = 0.0;
+    }
+    else q = q_out;
+    get_out_goal_.pose.orientation = q;
+  }
+
+  //if get_in_need
+  // check orientation ?
```
Comment (PhileasL) [inline src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp:132]: So, the response ?
No, assuming the goal orientation in specific area is in [0, 90, 180, 270]° orientation.

**Review by PhileasL — COMMENTED** [review]
- [ ] Check if a specific controller can use a specific goal_checker_plugin.
- [ ] Use `!defaultsvalues`
- [ ] Rename controllers
- [ ] Check dependencies
- [ ] Clean code (remove commentaries, logs, ...)

File: src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  return 10;
+}
+
+void SplitGoal::setAutoQuaternions(){
+  geometry_msgs::msg::Quaternion q;
+
+  if (get_out_need_){
+    geometry_msgs::msg::Quaternion q_out = current_pose_.pose.orientation;
+    RCLCPP_WARN(node_->get_logger(), "q_out z: %f, w: %f", q_out.z, q_out.w);
+    if ( (abs(q_out.z - 0.7071068)<0.22 &&  abs(q_out.w - 0.7071068)<0.22)
```
Comment (3wnbr1) [inline src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp:108]: Use a `const float var = 0.7071068` damn ! 😂
Same for `.22`

**Review by 3wnbr1 — CHANGES_REQUESTED** [review]
Amazing 👍

**Comment by PhileasL** [comment]
I'd like to be able to specify specific areas, exit area coordinates and distance from walls into a yml file.

- [ ] Check the feasibility

**Review by 3wnbr1 — APPROVED** [review]

File: src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp (hunk 6 above)
Comment (3wnbr1) [reply src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp:145]: @PhileasL just fix that one before merging

Comment (PhileasL) [reply src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp:145]: Yes, i'm not forgetting, i'm not done with dependencies either

Comment (PhileasL) [reply src/navigation/bt_custom_plugins/bt_split_goal/src/split_goal_action.cpp]: [bt_split_goal const float](https://github.com/robotique-ecam/cdfr/pull/36/commits/8280c7d8f247f57d076912db2d3bfb1d1a8c7068)

**Review by 3wnbr1 — APPROVED** [review]

**Pull request merged**